src/shogun/io/streaming/ParseBuffer.h

```diff
@@ -99,11 +99,14 @@ template <class T> class CParseBuffer: public CSGObject
 	Example<T>* get_free_example()
 	{
 #ifdef HAVE_CXX11
-		std::unique_lock<std::mutex> write_lk(*write_mutex, std::defer_lock);
-		std::unique_lock<std::mutex> current_ex_lock(*ex_in_use_mutex[ex_write_index], std::defer_lock);
-		std::lock(write_lk, current_ex_lock);
-		while (ex_used[ex_write_index] == E_NOT_USED)
-			ex_in_use_cond[ex_write_index]->wait(current_ex_lock);
+//		std::unique_lock<std::mutex> write_lk(*write_mutex, std::defer_lock);
+//		std::unique_lock<std::mutex> current_ex_lock(*ex_in_use_mutex[ex_write_index], std::defer_lock);
+//		std::lock(write_lk, current_ex_lock);
+        int32_t current_write_index = ex_write_index.load(std::memory_order_relaxed);  // only written from parser thread
+        int32_t current_read_index = ex_read_index.load(std::memory_order_acquire); // for synchronisation
+		while (ex_used[current_write_index] == E_NOT_USED)
```

Comment:
How do I profile it using valgrind (or something else)? 

I am thinking of making the elements in ex_used atomic(like `vector<atomic<E_IS_EXAMPLE_USED> > ex_used`) then we can use something like 
```C++
while(ex_used[current_index].exchange(E_NOT_USED)==E_NOT_USED){}
...//get the example
ex_used[current_index].store(E_USED);
```
If not can you elaborate more on how to use exchange for our case?